website/blog/2023-02-14-pgo.md

```diff
@@ -0,0 +1,101 @@
+- Convert the `.profraw` file into a `.profdata` file using LLVM's `llvm-profdata` tool.
+- Compile the program again, this time making use of the profiling data.
+
+## Preparation
+
+The data collected during the run will eventually need to be converted with `llvm-profdata`, which can be provided by installing the `llvm-tools-preview` component via `rustup`, or consider using recent versions of LLVM and the program provided by Clang.
+
+```bash
+rustup component add llvm-tools-preview
+```
+
+After installation `llvm-profdata` may need to be added to the `PATH` at the following path.
+
+```bash
+~/.rustup/toolchains/<toolchain>/lib/rustlib/<target-triple>/bin/
+```
+
+## Step-By-Step
+
+Rather than choosing a specific production environment workload, Databend's SQL logic tests are used here as a demonstration. It may not be positive in performance, but it can help us to understand how the process is carried out.
+
+> The caveat, however, is that the sample of data fed to the program during the profiling stage must be statistically representative of the typical usage scenarios; otherwise, profile-guided feedback has the potential to harm the overall performance of the final build instead of improving it.
+
+1. Make sure there is no left-over profiling data from previous runs.
+
+  ```bash
+  rm -rf /tmp/pgo-data
+  ```
+
+2. Build the instrumented binaries (with `release` profile), using the `RUSTFLAGS` environment variable in order to pass the PGO compiler flags to the compilation of all crates in the program.
+
+  ```bash
+  RUSTFLAGS="-Cprofile-generate=/tmp/pgo-data" \
+      cargo build --release --target=x86_64-unknown-linux-gnu
+  ```
+
+3. Instrumented binaries were run with some typical workload and we strongly recommend using workload that is statistically representative of the real scenario. This example runs SQL logic tests for reference only.
+  - Start a stand-alone Databend via a script, or a Databend cluster, given that production environments are more likely to run in cluster mode.
+  - Import the dataset and run a typical query workload.
+
+  ```bash
+  BUILD_PROFILE=release ./scripts/ci/deploy/databend-query-standalone.sh
+  ulimit -n 10000;ulimit -s 16384; cargo run -p sqllogictests --release -- --enable_sandbox --parallel 16 --no-fail-fast
+  ```
+
+4. Merge the `.profraw` files into a `.profdata` file with `llvm-profdata`.
+
+  ```bash
+  llvm-profdata merge -o /tmp/pgo-data/merged.profdata /tmp/pgo-data
+  ```
+
+5. Use the `.profdata` file for guiding optimizations. In fact, you can notice that both builds use the `--release` flag, because in the actual runtime case we always use the release build binary.
+
+  ```bash
+  RUSTFLAGS="-Cprofile-use=/tmp/pgo-data/merged.profdata -Cllvm-args=-pgo-warn-missing-function" \
+      cargo build --release --target=x86_64-unknown-linux-gnu
+  ```
+
+6. Run the compiled program again, running the previous workload to check performance.
```

Comment:
Run the compiled program again with the previous workload and check the performance: